crates/sui-graphql-rpc/src/types/balance.rs

```diff
@@ -110,37 +117,75 @@ impl Balance {
 
         Ok(conn)
     }
-}
 
-impl Target<Cursor> for StoredBalance {
-    type Source = objects::table;
+    fn base_query(address: SuiAddress, coin_type: Option<TypeTag>, lhs: i64, rhs: i64) -> RawQuery {
+        let mut snapshot_objs = query!(r#"SELECT * FROM objects_snapshot"#);
+        snapshot_objs = Balance::filter(snapshot_objs, address, coin_type.clone());
+
+        let mut history_objs = query!(r#"SELECT * FROM objects_history"#);
+        history_objs = Balance::filter(history_objs, address, coin_type.clone());
+        history_objs = filter!(
+            history_objs,
+            format!(r#"checkpoint_sequence_number BETWEEN {} AND {}"#, lhs, rhs)
+        );
+
+        let final_ = consistent_object_read(snapshot_objs, history_objs, lhs, rhs);
+        final_.group_by("candidates.coin_type")
+    }
 
-    fn filter_ge<ST, GB>(cursor: &Cursor, query: Query<ST, GB>) -> Query<ST, GB> {
-        query.filter(objects::dsl::coin_type.ge((**cursor).clone()))
+    fn filter(mut query: RawQuery, address: SuiAddress, coin_type: Option<TypeTag>) -> RawQuery {
+        query = filter!(
+            query,
+            format!(
+                "owner_id = '\\x{}'::bytea AND coin_type IS NOT NULL",
+                hex::encode(address.into_vec())
+            )
+        );
+
+        if let Some(coin_type) = coin_type {
+            query = filter!(
+                query,
+                "coin_type = {}",
+                coin_type.to_canonical_display(/* with_prefix */ true)
+            );
+        };
+
+        query
     }
+}
 
-    fn filter_le<ST, GB>(cursor: &Cursor, query: Query<ST, GB>) -> Query<ST, GB> {
-        query.filter(objects::dsl::coin_type.le((**cursor).clone()))
+impl RawPaginated<Cursor> for StoredBalance {
+    fn filter_ge(cursor: &Cursor, query: RawQuery) -> RawQuery {
+        filter!(query, "candidates.coin_type >= {}", (**cursor).clone())
```

Comment:
Probably worth adding a comment here explaining where "candidates" comes from.

Reply:
Although actually, do you need that at all? (e.g. could you get away with not specifying `candidates.`?

Reply:
Actually yeah, I can get away with not specifying `candidates`. To do so, we have to make sure `coin_type` is not a column on `newer`.

Reply:
```
                candidates.object_id = newer.object_id
                AND candidates.object_version < newer.object_version
```

Just this bit where we need both aliases. It'll only work for the balance query since the columns selected from `candidates` and `newer` do not overlap.